cmd/zrepl/mgmt_conn.c

```diff
@@ -860,21 +879,58 @@ process_message(uzfs_mgmt_conn_t *conn)
 		}
 		*snap++ = '\0';
 		/* ref will be released when async command has finished */
-		if (((zinfo = uzfs_zinfo_lookup(zvol_name)) == NULL) ||
-		    (zinfo->mgmt_conn != conn)) {
+		if ((zinfo = uzfs_zinfo_lookup(zvol_name)) == NULL) {
 			LOG_ERR("Unknown zvol: %s", zvol_name);
 			rc = reply_nodata(conn, ZVOL_OP_STATUS_FAILED,
 			    hdrp->opcode, hdrp->io_seq);
 			break;
 		}
+		if (zinfo->mgmt_conn != conn) {
+			uzfs_zinfo_drop_refcnt(zinfo, B_FALSE);
+			LOG_ERR("Target used invalid connection for "
+			    "zvol %s\n", zvol_name);
+			rc = reply_nodata(conn, ZVOL_OP_STATUS_FAILED,
+			    hdrp->opcode, hdrp->io_seq);
+			break;
+		}
 		if (hdrp->opcode == ZVOL_OPCODE_SNAP_CREATE) {
 			DBGCONN(conn, "Create snapshot command for %s@%s",
 			    zinfo->name, snap);
 		} else {
 			DBGCONN(conn, "Destroy snapshot command for %s@%s",
 			    zinfo->name, snap);
 		}
-		rc = uzfs_zvol_dispatch_command(conn, hdrp, snap, zinfo);
+		rc = uzfs_zvol_dispatch_command(conn, hdrp, snap,
+		    strlen(snap) + 1, zinfo);
+		break;
+
+	case ZVOL_OPCODE_RESIZE:
+		if (payload_size != sizeof (*resize_data)) {
+			rc = reply_nodata(conn, ZVOL_OP_STATUS_FAILED,
+			    hdrp->opcode, hdrp->io_seq);
+			break;
+		}
+		resize_data = payload;
+
+		/* ref will be released when async command has finished */
+		if ((zinfo = uzfs_zinfo_lookup(resize_data->volname)) == NULL) {
```

Comment:
having a function which does sanity check on conn will help?
not required for this PR.

Reply:
indeed there is too much repeating code. We will need to have a look at it in future.